Rework Key Update (#2237)
quicwg/base-drafts: draft-ietf-quic-transport.md

Initial[0]: CRYPTO[CH] ->

                                 Initial[0]: CRYPTO[SH] ACK[0]
           Handshake[0]: KEYS_READY, CRYPTO[EE, CERT, CV, FIN]
                                 <- 1-RTT[0]: STREAM[1, "..."]

> Initial[1]: ACK[0]

Reviewer: Is this Initial necessary? Assuming it is coalesced with the Handshake that contains KEYS_READY, wouldn't that cause the Initial keys to be thrown away, implicitly ACK'ing outstanding Initial packets?

Author: You are right.  The client might send this because it processes the Initial before it processes the KEYS_ACTIVE frame, but it doesn't have to send it.

Handshake[0]: KEYS_READY, CRYPTO[FIN], ACK[0]
1-RTT[0]: KEYS_READY, STREAM[0, "..."], ACK[0] ->

               1-RTT[1]: KEYS_READY, STREAM[55, "..."], ACK[0]

>                                        <- Handshake[1]: ACK[0]

Reviewer: Similarly with this Handshake packet. Is it necessary, assuming it was coalesced with the 1-RTT packet? On a side note, if we don't get rid of it, I think it might be better to put the Handshake packet before the 1-RTT packet in the list, since that's the order they'd be coalesced (ditto for the 0-RTT example below).

Author: Yeah, I'll remove it on the basis that it's a nice optimization.

Initial[0]: CRYPTO[CH] ->

                                 Initial[0]: CRYPTO[SH] ACK[0]
           Handshake[0]: KEYS_ACTIVE, CRYPTO[EE, CERT, CV, FIN]

>                                  <- 1-RTT[0]: STREAM[1, "..."]

Reviewer 2: Sending the KEY ACTIVE there seems wrong. The logic "I have processed a packet at this key level". That has not happened yet, will only happen at the sending of `<- Handshake[1]: ACK[0]`. That's the same reason why you correctly do not place a Key Active frame in the packet `<- 1-RTT[0]: STREAM[1, "..."]`

Reviewer: My understanding is that the logic to send a KEYS_ACTIVE frame is that you have both read and write keys available for use. Not that you have used them. For that reason, the server doesn't send the KEYS_ACTIVE for 1-RTT, but it does send it for Handshake.

0-RTT[0]: STREAM[0, "..."] ->

                                 Initial[0]: CRYPTO[SH] ACK[0]
           Handshake[0]: KEYS_ACTIVE, CRYPTO[EE, CERT, CV, FIN]

>                           <- 1-RTT[0]: STREAM[1, "..."] ACK[0]

Reviewer 2: Same issue. You cannot send Key Active there because the server has not yet received an Handshake packet from the peer.

* PING and PADDING frames contain no information, so lost PING or PADDING frames
  do not require repair.

* KEYS_ACTIVE frames are sent until acknowledged or until newer keys are used

>   for sending packets.

Reviewer 3: I'm not sure how this works. If KEYS_ACTIVE is not yet acked, we don't know if the peer has seen this KEYS_UPDATE YET.  We therefore cannot assume that the peer can accept another key update, and so this endpoint cannot use newer keys until an ack is received for the previously sent KEYS_UPDATE.

If this is correct, you should point this out earlier in "Responding to a Key Update" above as well.

Author: It is correct.

Interesting observation:  If your KEYS_ACTIVE got through and you don't see the ACK before your peer updates again, then you can just retransmit, but it's a different frame at that point.

That suggests that maybe you don't need to worry about stopping retransmission for KEYS_ACTIVE.  The sender can always update, and the sender of a packet with new keys containing KEYS_ACTIVE is always receptive to a key update.

Technically however, KEYS_ACTIVE sent in those circumstances has different semantics, so the text is correct as is.

the following sections, several long-header packets in this version of QUIC
contain these additional fields:

> Reserved Bit (R):

Reviewer 3: It's still two bits.

~~~
+-+-+-+-+-+-+-+-+
|1|1| 0 | R | P |

> +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

Reviewer 2: Unrelated change. Makes the PR harder to read.

frame sent in the first Initial packet.  Cryptographic handshake messages
subsequent to the first do not need to fit within a single UDP datagram.

#### Abandoning Initial Packets {#discard-initial}

Endpoints cease both sending and processing Initial packets when it both sends
and receives a Handshake or 1-RTT packet containing a KEYS_ACTIVE frame.  Though
packets might still be in flight or awaiting acknowledgment, no further Initial
packets need to be exchanged beyond this point.  Initial packet protection keys
are discarded (see Section 4.10 of {{QUIC-TLS}}) along with any loss recovery
and congestion control state (see Sections 5.3.1.2 and 6.9 of {{QUIC-RECOVERY}}).

> 

Reviewer 2: The server side specification feels wrong. The server does not know that the peer has received the server initial until it receives the KEY ACTIVE (Handshake) from the client. Yes, there is an "implicit" dependency that the Handshake cannot be processed without receiving first the Initial packet. There will be cases where Initial is lost, and both Initial and Handshake have to be retransmitted.

Reviewer: The KEYS_ACTIVE is sent when both read and write keys are ready, but you don't throw away the previous keys until you have sent and received a KEYS_ACTIVE. I'm unsure of what the problem is here. If you receive a KEYS_ACTIVE then the peer must have received all the necessary packets of yours sent with the previous key.

#### Abandoning Initial Packets {#discard-initial}

Endpoints cease both sending and processing Initial packets when it both sends

> and receives a Handshake packet containing a KEYS_ACTIVE frame.  Though packets

Reviewer 4: I think we should say "sends and receives a Handshake packet _or a 1-RTT packet_ containing a KEYS_ACTIVE frame", assuming that we wouldn't be waiting for ACKs when dropping the keys that are to carry the ACKs.

This is because a Handshake packet carrying just the KEYS_ACTIVE frame can be dropped, while all other packets reach the peers.

 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+
|0|1|S| R |K| P |

> +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

Reviewer 2: Unrelated change.

  This SHOULD be a UTF-8 encoded string {{!RFC3629}}.

## KEYS_ACTIVE Frame {#frame-keys-active}

An endpoint sends a KEYS_ACTIVE frame (type=0x1e) to signal that it has
installed keys for both sending and receiving packets.  Once this frame has been
both sent and received, older keys can be safely discarded.

The KEYS_ACTIVE frame contains no additional fields.

The packet that carries a KEYS_ACTIVE frame determines which keys are active and
usable.  The keys with the same key phase as those used in the packet that
carries the KEYS_ACTIVE frame are active.

> An endpoint MUST send a KEYS_ACTIVE packet in the first packet it sends using

Reviewer 3: ```suggestion
An endpoint MUST send a KEYS_ACTIVE frame in the first packet it sends using
```

> keys, but only if it is also able to receive packets that are protected using

Reviewer 3: ```suggestion
those keys, but only if it is also able to receive packets that are protected using
```

the corresponding keys.

KEYS_ACTIVE frames are retransmitted when declared lost, however implementations
need to take care not to retransmit lost KEYS_ACTIVE frames if they initiate a
subsequent key update.  This can happen if an acknowledgment for a packet

> containing a KEYS_ACTIVE frame is lost.

Reviewer 3: As I wrote above, I don't understand how this can happen.

Author: Say I initiate an update.  I receive a KEYS_ACTIVE in an updated packet.

At this point, I am expected to start sending KEYS_ACTIVE.  But I can also initiate another key update.

If I initiate a key update, which I can do without sending any more packets, then I MUST NOT send a KEYS_ACTIVE in that packet, because that leaves me open to a double update from my peer.

Endpoints MUST NOT send KEYS_ACTIVE frames in Initial or 0-RTT packets.

> 

Reviewer 3: I would move these last 3 paras to a new subsection called "Key Updates During the Handshake" in the "Key Update" section above.

A KEYS_ACTIVE frame used during the handshake can be used to indicate the
availability of Handshake keys by including it in a Handshake packet.  An
endpoint sends this frame in its first Handshake packet.  Once received, an
endpoint can discard Initial keys.

> A KEYS_ACTIVE frame used after the completion of the handshake in 1-RTT packets

Reviewer 3: ```suggestion
A KEYS_ACTIVE frame used in 1-RTT packets immediately after the completion of the handshake
```
(not any time after handshake but immediately after)

> indicates that Handshake keys are no longer needed.  A client sends this frame

Reviewer 3: ```suggestion
indicates that Handshake keys are no longer needed.  A client SHOULD send this frame
```

> in its first 1-RTT packet, and a server sends this frame in the first packet it

Reviewer 3: ```suggestion
in its first 1-RTT packet, and a server SHOULD send this frame in the first packet it
```

Author: I'm going with MUST.

> sends after completing the handshake.  A server might send 1-RTT keys prior to

Reviewer 3: A server never sends 1-RTT keys. I'm not sure what this sentence is trying to say.

Author: Oops, I meant packets.

  This SHOULD be a UTF-8 encoded string {{!RFC3629}}.

## KEYS_READY Frame {#frame-keys-ready}

An endpoint sends a KEYS_READY frame (type=0x1e) to signal that it has installed

> keys for reading and writing packets.  Receipt of this frame in a packet

Reviewer: ```suggestion
keys for both reading and writing packets.  Receipt of this frame in a packet
```
To be a bit more clear, it seems you are only supposed to send this when you are both keys are ready for use, not one or the other, correct?

Author: Yes, both.  And the clarification is good.

> indicates that all earlier keys can be safely discarded.

Reviewer 4: At least for dropping the Initial and Handshake keys, I had assumed that we would be sending the frame after all the transmissions have been acknowledged rather than when the new keys become available.

The concern of _not_ waiting for ACKs is that packets would be deemed lost, having negative effect on the congestion control. For example, the proposed design suggests the server to drop Handshake keys when it receives ClientFinished, without sending an ACK for the packet that carried ClientFinished.

Are we fine with that?

The other issue regarding the text is that it does not state that the signal is unilateral. My understanding is that an endpoint is expected to drop the keys when _both_ of the following conditions are met: 1) the endpoint has received the _done_ frame, 2) the endpoint is fine with dropping the keys. The text seems to not capture the second condition.

Author: We've had some discussions about losing packets during the handshake due to reordering, and I think that we just have to make some allowances in the congestion controller/loss recovery for that possibility.  For packets that we abandon like this, we lose the positive signal (i.e., congestion window increase), but we don't necessarily create a negative signal, because abandonment means that you don't declare them lost.

I'll tweak the text about the dropping of keys requiring both sending and receiving of the frame.

Author: Note the receiving only is a hangup from before I added the bit about delaying transmission of the frame in order to limit the number of active keys.  If you don't have that, receipt is sufficient because receiving the frame always implies that you need to send a matching frame.

Reviewer 4: > I think that we just have to make some allowances in the congestion controller/loss recovery for that possibility. For packets that we abandon like this, we lose the positive signal (i.e., congestion window increase), but we don't necessarily create a negative signal, because abandonment means that you don't declare them lost.

I think you are correct in pointing out that there's no negative signal even though we are losing a positive signal. As implied in my previous comment, my preference goes to delaying dropping keys until all the transmissions for that epoch is complete because that allows us to decouple loss-recovery / congestion control and epochs (FWIW, we will not bee seeing ACKs for ServerHello and ClientFinished). But I would not argue strongly, because there are ways to compensate for the missing signal.

> I'll tweak the text about the dropping of keys requiring both sending and receiving of the frame.

:+1:

Author: I don't want to delay discarding Initial keys, because we're would lose the DoS fix. Handshake keys are probably ok, but it is easier to draw the line between handshake and key update.

I don't expect that loss and reordering will be that significant, nor will the loss of positive signals. Any missing packets should be small (just an ACK probably), so the reduced increase to congestion allowance would be negligible.

We could add text to Section 6.2.2 of the recovery draft to this effect.

Reviewer 3: We should consider losses of handshake packets as not triggering congestion window reductions. This is perhaps a bit aggressive, but I would argue that we shouldn't have a lot of these packets anyways, and there are reasons beyond congestion that could cause these packets to be marked as lost.

I think we should go with the cleanest mechanism to drop keys -- this has been tricky to get right -- and we'll fix recovery around that. Once the other mechanisms are set, we can figure out how to best do recovery around them.

Reviewer 4: IMO the most architecturally sane way would be to transmit all necessary ACKs for every epoch, and I am not sure how much I buy the argument that dropping Initial keys earlier is a win, considering that there would be 1-RTT injection window anyways. A middlebox that observes an Initial can send CONNECTION_CLOSE to both endpoints (as they do for TCP), regardless of when we drop the Initial key. However I also agree that losing the ability to see the ACK is not a big deal. So maybe it's nothing more than a matter of taste.

Reviewer: I personally prefer the immediate discarding of keys and implicit ACKs.

Reviewer 5: @kazuho, but implementations could quite reasonably be skeptical of packets with CONNECTION_CLOSE and wait to see if others arrive.  Injecting a fictitious Initial packet with a broken/different handshake is a much more effective attack.  It's true that 1 RTT is probably plenty of time for an attacker, but I'd prefer to minimize the window anyway.
